File: clients/raycast/src/api/metadata.ts
import { getServerUrl } from "./config";

export interface Feed {
  url: string;
  title: string | null;
  format: "rss" | "atom" | "json_feed";
}

export interface MetadataFieldType {
  id: number;
  name: string;
}

export interface ClassifiedProfile {
  url: string;
  type: string;
}

export interface ContactMetadataResponse {
  name?: string;
  url?: string;
  email?: string;
  photo?: string;
  bio?: string;
  location?: string;
  feeds: Feed[];
  social_profiles: ClassifiedProfile[];
}

// Map profile type names to field type IDs (matching server enum)
const FIELD_TYPE_MAP: Record<string, MetadataFieldType> = {
  Bluesky: { id: 1, name: "Bluesky" },
  Email: { id: 2, name: "Email" },
  GitHub: { id: 3, name: "GitHub" },
  LinkedIn: { id: 4, name: "LinkedIn" },
  Mastodon: { id: 5, name: "Mastodon" },
  Website: { id: 6, name: "Website" },
  X: { id: 7, name: "X" },
  Other: { id: 8, name: "Other" },
};

export interface ClassifiedProfileWithFieldType {
  url: string;
  field_type: MetadataFieldType;
}

export function classifyProfile(profile: ClassifiedProfile): ClassifiedProfileWithFieldType {
  const fieldType = FIELD_TYPE_MAP[profile.type] || FIELD_TYPE_MAP["Other"];
  return { url: profile.url, field_type: fieldType };
}

export async function fetchContactMetadata(url: string): Promise<ContactMetadataResponse> {
  const params = new URLSearchParams({ url });
  const response = await fetch(`${getServerUrl()}/discovery/connection-metadata?${params.toString()}`);
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to fetch metadata");
  }
  return response.json();
}

File: clients/macos/Connections/Models/Metadata.swift
import Foundation

struct ClassifiedProfile: Codable, Hashable {
    let url: String
    let type: String
}

struct ContactMetadataResponse: Codable {
    let name: String?
    let url: String?
    let email: String?
    let photo: String?
    let bio: String?
    let location: String?
    let feeds: [DiscoveredFeed]
    let socialProfiles: [ClassifiedProfile]
}

struct DiscoveredFeed: Codable, Hashable {
    let url: String
    let title: String?
    let format: String
}

struct UriMetadataResponse: Codable {
    let title: String?
    let description: String?
    let image: String?
    let publishedAt: String?
    let authorName: String?
    let siteName: String?
    let canonicalUrl: String?
    let tags: [String]
    let contentType: String?
}

// Maps profile type names to field type IDs
enum ProfileClassifier {
    private static let fieldTypeMap: [String: MetadataFieldType] = [
        "Bluesky": MetadataFieldType(id: 1, name: "Bluesky"),
        "Email": MetadataFieldType(id: 2, name: "Email"),
        "GitHub": MetadataFieldType(id: 3, name: "GitHub"),
        "LinkedIn": MetadataFieldType(id: 4, name: "LinkedIn"),
        "Mastodon": MetadataFieldType(id: 5, name: "Mastodon"),
        "Website": MetadataFieldType(id: 6, name: "Website"),
        "X": MetadataFieldType(id: 7, name: "X"),
        "Other": MetadataFieldType(id: 8, name: "Other"),
    ]

    static func classify(_ profile: ClassifiedProfile) -> (url: String, fieldType: MetadataFieldType) {
        let fieldType = fieldTypeMap[profile.type] ?? fieldTypeMap["Other"]!
        return (profile.url, fieldType)
    }
}
